code/go/0chain.net/blobbercore/filestore/tree_validation.go

```diff
@@ -0,0 +1,436 @@
+// This file will validate challange tree and validation tree and provide merkle path for the provided
+// index/es as well.
+
+package filestore
+
+import (
+	"encoding/hex"
+	"errors"
+	"io"
+	"math"
+
+	"github.com/0chain/gosdk/core/util"
+	"golang.org/x/crypto/sha3"
+)
+
+const (
+	KB = 1024
+	MB = KB * KB
+)
+
+const (
+	HashSize = 32
+	// FMTSize is the size of tree nodes of fixed merkle tree excluding root node
+	FMTSize     = 65472
+	FMTDisKRead = 10 * MB
+)
+
+// fixedMerkleTree is used to verify fixed merkle tree root and store tree nodes to the file
+type fixedMerkleTree struct {
+	*util.FixedMerkleTree
+}
+
+// getNodesSize will calculate the size required to store tree nodes excluding root node.
+func getNodesSize(dataSize, merkleLeafSize int64) int64 {
+	totalLeaves := (dataSize + merkleLeafSize - 1) / merkleLeafSize
+	if totalLeaves == 1 {
+		return 0
+	}
+	totalNodes := totalLeaves
+	for totalLeaves > 2 {
+		totalLeaves = (totalLeaves + 1) / 2
+		totalNodes += totalLeaves
+
+	}
+	return totalNodes * HashSize
+}
+
+func calculateLeaves(dataSize int64) int {
+	return int(math.Ceil(float64(dataSize) / util.MaxMerkleLeavesSize))
+}
+
+func calculateDepth(totalLeaves int) int {
+	return int(math.Ceil(math.Log2(float64(totalLeaves)))) + 1
+}
+
+// CalculateRootAndStoreNodes will calculate all the intermediate nodes and write it to
+// f
+func (ft *fixedMerkleTree) CalculateRootAndStoreNodes(f io.Writer) (merkleRoot []byte, err error) {
+
+	nodes := make([][]byte, len(ft.Leaves))
+	for i := range ft.Leaves {
+		nodes[i] = ft.Leaves[i].GetHashBytes()
+	}
+
+	buffer := make([]byte, FMTSize)
+	var bufLen int
+	h := sha3.New256()
+
+	for i := 0; i < util.FixedMTDepth; i++ {
+		if len(nodes) == 1 {
+			break
+		}
+		newNodes := make([][]byte, (len(nodes)+1)/2)
+		var nodeInd int
+		if len(nodes)&1 == 1 {
+			nodes = append(nodes, nodes[len(nodes)-1])
+		}
+		for j := 0; j < len(nodes); j += 2 {
+			h.Reset()
+			prevBufLen := bufLen
+			bufLen += copy(buffer[bufLen:bufLen+HashSize], nodes[j])
+			bufLen += copy(buffer[bufLen:bufLen+HashSize], nodes[j+1])
+
+			h.Write(buffer[prevBufLen:bufLen])
+			newNodes[nodeInd] = h.Sum(nil)
+			nodeInd++
+		}
+
+		nodes = newNodes
+	}
+
+	_, err = f.Write(buffer)
+	if err != nil {
+		return nil, err
+	}
+
+	return nodes[0], nil
+}
+
+// fixedMerkleTreeProof is used to calculate merkle proof of certain index and also get the
+// content in batches.
+type fixedMerkleTreeProof struct {
+	idx      int
+	dataSize int64
+}
+
+func NewFMTPRoof(idx int, dataSize int64) *fixedMerkleTreeProof {
+	return &fixedMerkleTreeProof{
+		idx:      idx,
+		dataSize: dataSize,
+	}
+}
+
+// CalculateLeafContentLevelForIndex is used to calculate total levels or rows of a column of respective
```

Comment:
The data is not exact multiple of 64KB. Fixed merkle tree is based on this value where it will divide each 64KB into 1024 pieces.
Each such 1024 pieces are placed in each row.

Since we don't have zero padding anymore, in case of data that is not exact multiple of 64KB, end row will contain data upto some index only, say up to index 101. Now to fetch data for index 102, the `level_of_row` = `level_of_end_row` - 1 

Check https://github.com/0chain/blobber/wiki/File-Upload-And-Merkle-Root-Calculation out for reference.